src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,608 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.ConvertLinq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq
+{
+    internal sealed class CSharpConvertLinqQueryToLinqMethodProvider : AbstractConvertLinqProvider
+    {
+        private static readonly TypeSyntax VarNameIdentifier = SyntaxFactory.IdentifierName("var");
+
+        protected override IAnalyzer CreateAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+            => new CSharpAnalyzer(semanticModel, cancellationToken);
+
+        private sealed class CSharpAnalyzer : AnalyzerBase<QueryExpressionSyntax, SyntaxNode, SyntaxNode>
+        {
+            public CSharpAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+                : base(semanticModel, cancellationToken)
+            {
+            }
+
+            protected override string Title => CSharpFeaturesResources.Convert_LINQ_query_to_foreach;
+
+            protected override bool TryConvert(QueryExpressionSyntax source, out DocumentUpdate documentUpdate)
+            {
+                if (_semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any() ||
```

Comment:
>_semanticModel.GetDiagnostics(source.Span, _cancellationToken) [](start = 20, length = 62)

`GetDiagnostics` call is actually very expensive. Can we instead just have syntactic error check `source.HasErrors`? If the node meets all syntax criterias following it then the last step for `TryConvert` can perform GetDiagnostics check.

Reply:
i.e.

```
if (!TryConvertInternal(..., out var result) || GetDiagnostics(...).Any())
{
   result = null;
   return false;
}

return true;
```

---
In reply to: [176519698](https://github.com/dotnet/roslyn/pull/25362#discussion_r176519698) [](ancestors = 176519698)

Reply:
This one checks for diagnostics of the original QueryExpression node (not the whole document).

documentUpdate(Info) is a container for a node to be replaced with multiple node replacing it.
I think we can check diagnostics after only when replaced the old node (it is not a QueryExpresison but the parent statement).

Do you mean to call GetDiagnostics for the parent for the nodes replacing the statement?

---
In reply to: [176520283](https://github.com/dotnet/roslyn/pull/25362#discussion_r176520283) [](ancestors = 176520283,176519698)